orttraining/orttraining/python/training/checkpoint.py

```diff
@@ -327,28 +377,107 @@ def aggregate_checkpoints(paths, pytorch_format=True):
                 "Optimizer name mismatch among checkpoint files. File: {}".format(path)
 
         # aggregate all model states
-        _aggregate_model_states(rank_state_dict, sharded_states_original_dims, state_dict, loaded_mixed_precision)
+        _aggregate_model_states(rank_state_dict, sharded_states_original_dims, state_dict, loaded_mixed_precision, mode)
 
         if not pytorch_format:
             # aggregate all optimizer states if pytorch_format is False
-            _aggregate_optimizer_states(rank_state_dict, sharded_states_original_dims, state_dict)
+            _aggregate_optimizer_states(rank_state_dict, sharded_states_original_dims, state_dict, mode)
+
+            # for D+H aggregation scenario, the first pass of aggregation(partial aggregation) is over D groups 
+            # to aggregate over Zero, and another pass to aggregate Megatron partitioned
+            # states. Preserve the relevant partition info only for weights that are megatron partitioned for 
+            # a partial aggregation call
+            if partial_aggregation:
+                _aggregate_megatron_partition_info(rank_state_dict, state_dict)
 
             # entry for trainer_options in the state_dict to perform other sanity checks
             if _utils.state_dict_trainer_options_key() not in state_dict:
-                _aggregate_trainer_options(rank_state_dict, state_dict)
+                _aggregate_trainer_options(rank_state_dict, state_dict, partial_aggregation)
 
             # entry for user_dict in the state_dict if not already present
             if _utils.state_dict_user_dict_key() not in state_dict and \
                 _utils.state_dict_user_dict_key() in rank_state_dict:
                 state_dict[_utils.state_dict_user_dict_key()] = rank_state_dict[_utils.state_dict_user_dict_key()]
 
-    # reshape all the sharded tensors based on the original dimensions stored in sharded_states_original_dims
-    _reshape_states(sharded_states_original_dims, state_dict, loaded_mixed_precision)
+    # for a partial aggregation scenario, we might not have the entire tensor aggregated yet, thus skip reshape
+    if not partial_aggregation:
+        # reshape all the sharded tensors based on the original dimensions stored in sharded_states_original_dims
+        _reshape_states(sharded_states_original_dims, state_dict, loaded_mixed_precision)
 
     # return a flat structure for PyTorch model in case pytorch_format is True
     # else return the hierarchical structure for ORTTrainer
     return _to_pytorch_format(state_dict) if pytorch_format else state_dict
 
+def _aggregate_over_D_H(ordered_paths, D_groups, H_groups, pytorch_format):
+    sharded_states_original_dims = {}
+    save_dir = tempfile.mkdtemp()
```

Comment:
one small comment here. May we use a higher level function that auto cleans up after tempfile has been closed such as ```tempfile.TemporaryFile``` instead of using ```tempfile.mkdtemp```? We could also consider using ```with``` in that case.

Reply:
giving tempfile.TemporaryDirectory() a shot instead for automatic cleanup of the temp dir created